include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	 */
+	// Not possible when working with references, maybe only for actual node parameters?
+	// std::shared_ptr<Parameter<ParameterT>> get(const std::string& name);
+	// std::shared_ptr<const Parameter<ParameterT>> get(const std::string& name) const noexcept;
+
+	//
+	// TODO: insert utility functions:  validate(), print, ostream<<, errors(), warnings()...
+	//
+
+private:
+	const rclcpp::Node::SharedPtr node_;
+	const std::string base_name_;
+
+	// Class similar to the one in rosparam_shortcuts/node_parameters.h
+	// implements ROS calls/checks/callbacks for node_parameters_
+	NodeParametersInterface node_parameters_interface_;
+
+	// Store parameters, only node parameters (rclcpp::Parameters) are stored inside NodeConfig
+	std::map<std::string, Parameter&> parameters_;
+	std::map<std::string, std::shared_ptr<NodeParameter>> node_parameters_;
+
+	// Needed? keep track of registered callbacks that are required for linking SetParameters requests or
+	// exposing high-level update changes (invalid/valid states)
+	// std::string<std::string, std::string> registered_node_parameter_callbacks_ids_;
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+	ExampleConfig(const rclcpp::Node::SharedPtr& node) : NodeConfig(node) {
+		// declare simple parameters just like that
+		declare(int1_param);
+		declare(int2_param);
+		declare(int3_param);
+
+		// "double_param", add some description and validity checks
+		declare(double_param)
+		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
+		    .warnIf([](const double& val, std::string& message) {
+			    message = "Value is 7, are you sure this is a good value?";
+			    return val == 7;
+		    })
+		    .errorIf([](const double& val, std::string& message) {
+			    message = "Value must not be negative!";
+			    return val < 0;
+		    });
+
+		// "string_param", use a parameter error condition "EMPTY"
+		declare(string_param)
+		    .describe("meh")
+		    .warnIf([](const std::string& val, std::string& message) {
+			    message = "string is much too long";
+			    return val.size() > 100;
+		    })
+		    .errorIf(EMPTY);
+
+		// "double_map_param", validate keys and values, reject if not valid
+		declare(double_map_param).describe("6-dof joint state").errorIf([](const auto& values, std::string& message) {
+			message = "Invalid joint state map values, expected 6-dof";
+			return values.size() != 6 || checkJointNamesInRobotModel(values);  // from somewhere
+		});
+
+		//
+		// can already check for errors/warnings here...
+		//
+	}
+
+	Parameter<double> double_param{ "double_param" };
+	Parameter<std::vector<std::string>> string_param{ "string_param" };
+	Parameter<std::map<std::string, double>> double_map_param{ "double_map_param" };
+	Parameter<int> int1_param{ "int1_param" };
+	Parameter<int> int2_param{ "int2_param" };
+	Parameter<int> int3_param{ "int3_param" };
```

Comment:
I take it these are the keys of the parameters.  Would it make sense to have the keys supplied as a parameter to the declare function instead of here?  I'm just trying to get all the meta-data about each parameter declared in one place.

Reply:
Yes, that would make sense. We just need to make sure that it's not possible to change the name and to declare the parameter a second time.

Reply:
Is it really worth going for the explicit parameter members instead of `any` datatypes collected in a map?
It's somewhat cumbersome due to the repeated naming (unless you add pre-processor magic to repeat it for you) and you can't define the datatype, name, and additional data in the same command.
I'd prefer something like
```
declare<int>("parameter")
   .describe("my parameter")
   .errorIf("no prime number", [](auto x){ for(int i= 2; i < x; ++i) if(x%i == 0) return true; return false; });

Reply:
I thought about this as well, but to me the explicit members make the config API much more usable since it doesn't require type casting and accessing the map using (const?) keys which might also fail. Actually, the current API doesn't even prevent you from using `any` and a member map as long as you specify the corresponding parameter template.

Reply:
It's clearly two alternatives with their separate pros and cons.
There might be a middle-way using variadic templates, but even if there is, it would add a lot of complexity and expertise to get it right.

Type-casting is somewhat annoying indeed, but so is having the type declared separate from the actual declaration of everything.

One idea to move all information into one place with your current approach could be a best-practice using uniform initialization. Something like
```c++
Parameter<int> int1_param{
		name("int1_param").
		.describe("This is a very important double parameter", "should be greater than 0 and not 7")
		.warnIf([](const double& val, std::string& message) {
			message = "Value is 7, are you sure this is a good value?";
			return val == 7;
		})
		.errorIf([](const double& val, std::string& message) {
			message = "Value must not be negative!";
			return val < 0;
		}) };
```

That should be possible if you allow Parameter construction only from the builder-types.